### PR TITLE
Update Mac packaging to work without platypus

### LINKED ISSUE #18518: OS X .app not building correctly
There is a bug in platypus 5.1 which means that julia isn't actually being included in the .app:
https://github.com/sveinbjornt/Platypus/issues/78

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #18518

### CODE REVIEW

**Comment by simonbyrne** [comment]
cc @tkelman @staticfloat

File: contrib/mac/app/Makefile
Comment (simonbyrne) [inline contrib/mac/app/Makefile]: Can we get rid of this line?

**Comment by simonbyrne** [comment]
I've also got rid of some of the old cruft (like references to Winston and Cairo).

File: contrib/mac/app/Makefile
Comment (simonbyrne) [inline contrib/mac/app/Makefile]: @ViralBShah this was your contribution: is this still necessary? Do we still build .a or .la files?

**Comment by slarew** [comment]
@simonbyrne I'll look over this tonight when I have some free time.

File: contrib/mac/app/Makefile
Comment (tkelman) [inline contrib/mac/app/Makefile]: they get built by deps and we don't want to install them

Comment (simonbyrne) [inline contrib/mac/app/Makefile]: ok, I didn't see any in my build, that's why I was wondering.

Comment (tkelman) [inline contrib/mac/app/Makefile]: oh there might not be any under lib/julia at the moment

File: contrib/mac/app/Makefile
Comment (slarew) [inline contrib/mac/app/Makefile]: [CFBundleVersion](https://developer.apple.com/library/content/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/20001431-102364): "build version number should be a string comprised of three non-negative, period-separated integers with the first integer being greater than zero—for example, 3.1.2" ... "While developing a new version of your app, you can include a suffix after the number that is being updated; for example 3.1.3a1"

I believe the `JULIA_VERSION_OPT_COMMIT` variable does not always precisely follow these rules.  Maybe it doesn't matter too much in the case of nightlies which are not tagged commits.

File: contrib/mac/app/README.md
Comment (slarew) [inline contrib/mac/app/README.md]: OS X -> macOS

File: contrib/mac/app/Makefile
Comment (slarew) [inline contrib/mac/app/Makefile]: I believe strict adherence to [code signing rules](https://developer.apple.com/library/content/technotes/tn2206/_index.html) would disallow putting executable code (e.g. julia executable and deps libraries) in the Resources directory.  But this is a whole different issue...

**Review by slarew — COMMENTED** [review]
I haven't tried this out, but the changes look good to me.  I didn't know that osacompile could output an app bundle!  That simplifies it all greatly.

Might consider making sure all the [recommended Info.plist keys](https://developer.apple.com/library/content/documentation/General/Reference/InfoPlistKeyReference/Articles/AboutInformationPropertyListFiles.html#//apple_ref/doc/uid/TP40009254-SW6) are added.

File: contrib/mac/app/Makefile
Comment (slarew) [inline contrib/mac/app/Makefile]: Nevermind, [Wikipedia](https://en.wikipedia.org/wiki/Copyright#Copyright_notice) says just using the symbol should be fine.

File: contrib/mac/app/Makefile
Comment (simonbyrne) [inline contrib/mac/app/Makefile]: It should always be of the form `X.Y.Z`, `X.Y.Z-aaa` or `X.Y.Z-aaa-hash`. At the moment the first integer is zero, but there's not much we can do about that.

For reference, platypus would use `X.Y.Z` or `X.Y.Z-aaa`.

File: contrib/mac/app/Makefile
Comment (simonbyrne) [inline contrib/mac/app/Makefile]: I don't think so. The background for this is #9010, and as I said in that issue it's a pretty weird edge case which most apps don't bother with.  Maybe we should take this opportunity to get rid of it.

File: contrib/mac/app/Makefile
Comment (simonbyrne) [inline contrib/mac/app/Makefile]: Me either. It will be of that form for release versions, but for nightlies it can be `0.5.0-dev`

File: contrib/mac/app/Makefile
Comment (ViralBShah) [inline contrib/mac/app/Makefile]: I don't think this line is necessary anymore. I recently checked and noticed that we do not install the .a files.

File: contrib/mac/app/Makefile
Comment (simonbyrne) [inline contrib/mac/app/Makefile]: I had just recently changed it for this reason (it used to be just `org.julialang`), but we can be more specific here.

Does this mean that we can't later switch the code signer for the same ID?

Comment (slarew) [inline contrib/mac/app/Makefile]: AFAIK, as of now, there are two points where the code signature comes into play:
1. GateKeeper checks the identity of the signer, which by default must be Apple or a registered Developer ID.  If the signing identity doesn't pass GateKeeper checks, the code isn't allowed to execute.  So as long as the signing identity is a Developer ID or Apple, it can be changed in the future without affecting the default behavior.
2. The macOS application (network) firewall records its rules base on the code signature.  I have not tested this, but I think that if the signing identity changes, then the firewall rules may need to be updated. This is probably not a big deal.

File: contrib/mac/app/README.md
Comment (ararslan) [inline contrib/mac/app/README.md]: Technically it's still OS X up to Sierra, which is still unreleased

Comment (slarew) [inline contrib/mac/app/README.md]: I wonder if Apple provides official marketing guidelines on the OS X vs macOS naming issue.

File: contrib/mac/app/Makefile
Comment (tkelman) [inline contrib/mac/app/Makefile]: I misread that path initially, but yeah we should be removing static libs and libtool junk during `make install` and/or `make binary-dist` which is before it gets here

File: contrib/mac/app/Makefile
Comment (tkelman) [inline contrib/mac/app/Makefile]: are these created in the tarball now?

File: contrib/mac/app/Makefile
Comment (staticfloat) [inline contrib/mac/app/Makefile]: I say we see if we can get rid of it as we iterate on this PR.  I'd like to simplify as much as possible, and the spirit of #9010 was [the `chown` on line 30](https://github.com/JuliaLang/julia/blob/399710ea7dab478ffe85e3e898e8b6a85e45740e/contrib/mac/app/Makefile#L29-L30) anyway.

File: contrib/mac/app/Makefile
Comment (staticfloat) [inline contrib/mac/app/Makefile]: According to the updated [`Resource Rules`](https://developer.apple.com/library/content/technotes/tn2206/_index.html#//apple_ref/doc/uid/DTS40007919-CH1-TNTAG206), it looks like all files are signed, there is no distinction made between "Resources" and "Binaries" anymore, so the "security" side of this distinction should now be moot.

As far as adherence to protocol goes, [Apple states](https://developer.apple.com/library/content/technotes/tn2206/_index.html#//apple_ref/doc/uid/DTS40007919-CH1-TNTAG201):

> ...putting code into other places will cause it to be sealed as data (resource) files. This causes this code to be sealed twice; once in the nested code, and once in the outer signature. This wastes both signing and verification time and storage space. Also, this can break the outer signature of apps that use their own update mechanisms to replace nested code. If this nested code is being treated as resources, the outer signature doesn't know that this nested content is actually code.

> Always put code and data into their proper places. This applies to all signed code and is enforced by the code signing machinery regardless of how the code is distributed.

They don't threaten us with codesigning breaking if we don't follow the rules, but they do say that the Mac App Store may reject submissions that don't follow these rules without warning:

> Note: While strict compliance with these rules may not affect your app today, anything that doesn't meet these requirements note may be rejected by code signing verification (and the Mac App Store validator, in the case of Mac App Store apps) at any point in the future without notice.

All this being said, I'm not at all against moving things around, so if you have a better layout in mind I'd love to hear it.  We can move things around pretty flexibly now by setting make flags [such as these](https://github.com/JuliaLang/julia/blob/master/Make.inc#L176-L184) to place our binaries in the proper location, and our non-binaries in others.

File: contrib/mac/app/Makefile
Comment (simonbyrne) [inline contrib/mac/app/Makefile]: I've removed this now.

File: contrib/mac/app/Makefile
Comment (simonbyrne) [inline contrib/mac/app/Makefile]: This should be fixed now.

File: contrib/mac/app/Makefile
Comment (simonbyrne) [inline contrib/mac/app/Makefile]: They seem to be. Also, I think these were leftover from the days when we included Tk and Cairo libs.

File: contrib/mac/app/Makefile
Comment (simonbyrne) [inline contrib/mac/app/Makefile]: The existing structure is just the same as what we did with platypus, but there's no real reason we can't change it. As a first step, it would probably make sense to structure it as a app-local framework.

I don't think we need to move individual files around though. The R and Python frameworks both just keep the usual unix directory structure buried somewhere, and then symlink everything as necessary.

Comment (simonbyrne) [inline contrib/mac/app/Makefile]: That said, maybe this would be better as a different pull request.

**Comment by simonbyrne** [comment]
I'm fairly happy with this now. I think we can leave the framework structure for another issue.

**Comment by staticfloat** [comment]
I'm running the tip of this branch [through the buildbots](https://build.julialang.org/builders/package_osx10.9-x64/builds/857) to see what it comes up with.

**Comment by staticfloat** [comment]
Looks to work just fine, here's [the resultant binary](https://s3.amazonaws.com/julianightlies/bin/osx/x64/0.6/julia-0.6.0-aeadb11ecf-osx.dmg), grab it while it's hot!

**Review by staticfloat — APPROVED** [review]

**Comment by ViralBShah** [comment]
Binary seems to work well.

**Comment by tkelman** [comment]
Will this change the internal structure of the dmg in any way that we'd need to worry about on Travis for `language: julia`?

**Comment by staticfloat** [comment]
It shouldn't, from what I can tell.

**Comment by simonbyrne** [comment]
It shouldn't, I intentionally kept the directory structure the same.

**Comment by simonbyrne** [comment]
@tkelman I'll leave this up to you to pull the trigger on.

**Comment by tkelman** [comment]
I don't use or care about supporting this OS, but ok.

**Comment by ViralBShah** [comment]
Should we backport to 0.5?

**Comment by simonbyrne** [comment]
We can do, as it should be pretty much identical for users, but I'll leave this call to the release maintainers.

**Comment by tkelman** [comment]
if it doesn't cause any issues on the nightlies, it would be good for future-proofing to avoid accidentally upgrading to broken versions of things and releasing broken builds again

**Pull request merged**
